### DISCUSSION TITLE
Generate Conversions Manually when Needed

### DISCUSSION THREAD

**Discussion by danFWD** [spatie/laravel-medialibrary#1980] [Ideas] [answered]
I understand why generating conversions automatically when added to a model is a good idea, however, I have a scenario where it isn’t the preferred method. I am importing hundreds of thousands of images from a legacy app, and not only would it take forever to generate conversions for all of them, it would use up a lot more disk space. Since most of the images might never even be viewed again before being archived down the road, it only makes sense to generate conversions as needed.

Is it possible to add two methods to the Media class, one to disable auto-generation when adding a media item, and one to generate a conversion manually? Something like:

`  $myModel->addMedia( $img )->skipConversions()->etc;`
`  $myModel->getMedia('images')[0]->generateConversions(['thumb'], false);`

The `generateConversions` method would only generate a conversion if one doesn’t exist or the second argument is TRUE (force generate). Or possibly create a third method that `->generateConversionsIfNoneExist(['thumb']);`.

I really appreciate it. I am using v8 of the Media library.

**Comment by karelbyte** [comment]
Hello, did you find any way to do it? I also occupy something like that

**Reply by danFWD** [reply]
Still waiting for either an answer on StackOverflow or some engagement here. Otherwise, I might have to attempt to extend the class myself. Thx

**Comment by vesper8** [comment]
Did you ever get anywhere with this @danFWD ?

I need to do the same thing.. but in my case I also need to generate the conversion on the fly with `withManipulations` so I can pass in custom crop parameters

**Reply by danFWD** [reply]
No, after two years I haven’t heard anything on this. I am getting close to the end of beta and will need to do the import soon. I will have to import with conversions disabled, so my plan is either one of two things:

1. If an image is viewed but doesn't have a conversion, I will just clone it (with conversion enabled) and delete the old record. 

2. The library has a CLI command to regenerate images ([see docs](https://spatie.be/docs/laravel-medialibrary/v10/converting-images/regenerating-images)). I can write a console command to regenerate images using specific IDs ([see Laravel docs](https://laravel.com/docs/9.x/artisan#programmatically-executing-commands#programmatically-executing-commands)).

Number 2 seems like the more “Laravel” way to do it, however, it’s weird that there are methods to accomplish this within the codebase, but they aren’t exposed to the API - only the command line. I might need to dig a bit into those files and see if there is a hidden interface somewhere.

**Comment by poldixd** [comment]
I found a good tip on issue #1696. You can temporary empty the `image_generators` array in the config.

**Comment by danFWD** [comment, marked as answer]
Here is what I finally uncovered, though it still doesn't solve turning off the conversion when saving an image.

```PHP
/**
 * This checks to make sure the thumb conversion exists, 
 * and if not, goes ahead and creates it. 
 */
$fileManipulator = new \Spatie\MediaLibrary\Conversions\FileManipulator;
$media = $myModel->getMedia('images');
foreach( $media as $img ) {
 if( !$img->hasGeneratedConversion('thumb') ) {
  $fileManipulator->createDerivedFiles($img);
 }
}
```
Tested and this generates the conversion and saves the model conversion.

**Comment by danFWD** [comment]
Actually... upon thinking about this a bit deeper, if I create a static variable on my model and use that to turn on and off the registering of a conversion, like so:

```PHP
class MyModel extends Model implements HasMedia {
 use \Spatie\MediaLibrary\InteractsWithMedia;

 public static $conversionsEnabled = true;

 public function registerMediaConversions(Media $media = null): void {
  if( self::$conversionsEnabled ) {
    $this->addMediaConversion('thumb')
      ->width(420)
      ->height(265)
      ->quality(35)
      ->fit('crop', 420, 265)
      ->performOnCollections('images');
   }
 }

 // etc
}

// my script
MyModel::$conversionsEnabled = false;
 // import images
MyModel::$conversionsEnabled = true; // reset it
```
